### PR TITLE
feat(contracts): Refactor voting architecture to separate weight calculation from vote tracking

### PR DESCRIPTION
**Pull request by adamgall**
Depends on https://github.com/decentdao/decent-contracts/pull/415

Fixes [ENG-1266](https://linear.app/decent-labs/issue/ENG-1266/refactor-strategyvotingadapter-architecture-to-separate-voting-weight)

## Motivation

The existing VotingAdapter architecture had critical issues including tight coupling to IStrategyV1, mixed concerns between weight calculation and vote tracking, circular dependencies, and code duplication. This refactoring separates these concerns to create a more modular, maintainable, and extensible voting system.

## Change Summary

- Created new interfaces IVotingWeightStrategyV1 and IVoteTrackerV1 with clear separation of concerns
- Implemented VotingWeightERC20V1 and VotingWeightERC721V1 for weight calculation
- Implemented VoteTrackerERC20V1 and VoteTrackerERC721V1 for vote tracking
- Updated StrategyV1 to use new VotingConfig struct pairing weight strategy with vote tracker
- Updated FreezeVotingAzoriusV1 and SystemDeployerV1 to use new architecture
- Removed old VotingAdapter contracts and interfaces
- Fixed gasless voting for ERC-4337 paymasters with new getVotingWeightForPaymaster()
- Added comprehensive test coverage (88 tests) for all new contracts

# Project Plan

## Overview

This document outlines the complete refactoring of the Strategy/VotingAdapter architecture to separate voting weight calculation from vote tracking, creating a more modular and maintainable system.

## Problem Statement

The existing VotingAdapter architecture had several critical issues:

1. **Tight Coupling**: VotingAdapters were tightly coupled to IStrategyV1, making them unusable for other voting contexts
2. **Mixed Concerns**: VotingAdapters handled both weight calculation AND vote tracking in a single contract
3. **Circular Dependencies**: VotingAdapters depended on Strategy which depended on VotingAdapters
4. **Code Duplication**: Freeze voting logic was duplicated across different adapter types
5. **Inflexibility**: Hard to add new voting mechanisms without creating entirely new adapter contracts

## Proposed Architecture

The refactoring separated voting weight calculation from vote tracking:

```
User
 ↓
Strategy/FreezeVoting (Orchestration)
 ├→ IVotingWeightStrategy (Weight Calculation)
 └→ IVoteTracker (Vote Tracking)
```

### Why This Split?

The old VotingAdapter contracts were monolithic - they handled both:

1. **Calculating voting weight** (how many votes you have based on token holdings)
2. **Tracking votes** (preventing the same tokens from voting twice)

This created tight coupling and made the contracts inflexible. The new architecture splits these into two separate concerns that work together through composition.

### 1. New Core Interfaces (with V1 suffix)

**IVotingWeightStrategyV1**

- Calculates voting weight based on token holdings
- Processes vote data (e.g., NFT token IDs)
- Returns weight and processed data for tracking

**IVoteTrackerV1**

- Tracks votes to prevent double voting
- Context-aware (proposal ID based)
- Handles both address-based and token ID-based tracking

### 2. Concrete Implementations

**Voting Weight Strategies:**

- `VotingWeightERC20V1`: Calculates weight from ERC20 token balance at snapshot
- `VotingWeightERC721V1`: Calculates weight from NFT ownership with ID validation

**Vote Trackers:**

- `VoteTrackerERC20V1`: One vote per address per proposal
- `VoteTrackerERC721V1`: Tracks individual NFT usage to prevent reuse

### 3. Core Contract Updates

**StrategyV1 Changes:**

- Replaced `VotingAdapterVoteData` with `VotingConfigVoteData`
- Added `VotingConfig` struct that pairs weight calculation with vote tracking: ```solidity struct VotingConfig { address weightStrategy; // How to calculate voting power address voteTracker; // How to track votes } ```
- Updated `castVote` to use two-step process:

  ```solidity // 1. Calculate voting weight (read-only) (uint256 weight, bytes memory processedData) = IVotingWeightStrategyV1(config.weightStrategy).calculateWeight( voter, proposalStartTimestamp, voteData );

  // 2. Record the vote (state change) IVoteTrackerV1(config.voteTracker).recordVote( proposalId, voter, processedData ); ```

**FreezeVotingAzoriusV1 Updates:**

- Updated to use `VotingConfigVoteData`
- Maintained all freeze voting functionality

**SystemDeployerV1 Updates:**

- Replaced `VotingAdapterParams` with `VotingConfigParams`
- Deploy weight strategies and vote trackers separately
- Create `VotingConfig` structs for initialization

## How the New Architecture Works Together

### The Composition Pattern

The key innovation is the `VotingConfig` struct that pairs a weight strategy with a vote tracker. This enables flexible composition:

- **ERC20 Voting**: ERC20 weight strategy + ERC20 vote tracker
- **NFT Voting**: ERC721 weight strategy + ERC721 vote tracker
- **Future**: Could mix and match for novel mechanisms (e.g., ERC20 weight with NFT-style tracking)

### Context-Based Vote Tracking

Vote trackers use a "context ID" to separate different voting scenarios:

- **Normal proposals**: Use `proposalId` as context
- **Freeze voting**: Use `freezeProposalCreated` timestamp as context

This allows the same tokens to participate in both normal and freeze voting without conflict.

### Security Through Authorization

Vote trackers implement an authorization system:

- Only the Strategy contract can record votes
- Prevents unauthorized contracts from manipulating vote state
- Clear security boundary between read and write operations

## Technical Achievements

### 1. EIP-7201 Storage Patterns

All new contracts use namespaced storage for upgradeability safety:

- VotingWeightERC20V1: `0x4d0e0752be0f6c9ed5c28b9ff74819e1ab4b3a817e24bbdd087f60b66a2a4300`
- VotingWeightERC721V1: `0x5e4d3c2b1a0f9e8d7c6b5a4f3e2d1c0b9a8f7e6d5c4b3a2f1e0d9c8b7a6f5e00`
- VoteTrackerERC20V1: `0x2f3b5de5aa7c8e9e3a8c1f5e8e1d4c6f7a2b3c4d5e6f7a8b9c0d1e2f3a4b5c00`
- VoteTrackerERC721V1: `0x3a6b5c4d3e2f1a0b9c8d7e6f5a4b3c2d1e0f9a8b7c6d5e4f3a2b1c0d9e8f7a00`

### 2. Clean Architecture Benefits

- **Separation of Concerns**: Weight calculation is pure logic (view functions), vote tracking is isolated state management
- **Reusability**: Both normal and freeze voting use the same components
- **Testability**: Each component can be tested in isolation
- **Extensibility**: Easy to add new token types or voting mechanisms

### 3. Gasless Voting Fix for ERC-4337

During the refactoring, we discovered and fixed a critical issue with gasless voting validation:

- **Problem**: The old VotingAdapters used `getPastVotes()` which calls banned opcodes (block.timestamp, block.number) during ERC-4337 paymaster validation
- **Solution**: Added new `getVotingWeightForPaymaster()` function that manually iterates checkpoints for ERC20 tokens
- **Benefits**: Enables gasless voting through Light Accounts without triggering validation failures
- **Implementation**: Updated `validStrategyVote` to use the new paymaster-safe function

### 4. Migration Path

- No breaking changes to external interfaces
- Backwards compatibility maintained
- Old VotingAdapter contracts removed after review

## Future Extensions Enabled

This architecture now supports:

- Quadratic voting strategies
- Time-weighted voting
- Cross-chain voting via oracles
- Delegated voting patterns
- Multi-token voting in single proposal
- Standalone freeze voting mechanisms (next project)

## Status Summary

### Completed Work

- ✅ Created all new voting interfaces with V1 suffix
- ✅ Implemented all voting weight strategies
- ✅ Implemented all vote trackers
- ✅ Updated StrategyV1 to new architecture
- ✅ Updated FreezeVotingAzoriusV1
- ✅ Updated SystemDeployerV1
- ✅ Removed old VotingAdapter contracts
- ✅ Removed metadata field from VotingConfig struct
- ✅ Fixed gasless voting validation for ERC-4337 paymasters
- ✅ Added proper authorization to vote trackers
- ✅ Fixed interface consistency (VotingWeightERC20V1 now inherits from IVotingWeightERC20V1)
- ✅ Moved VotingConfigVoteData to separate IVotingTypes interface
- ✅ Updated SystemDeployerV1 to authorize strategy on vote trackers
- ✅ All contracts compile successfully
- ✅ All linting passes
- ✅ Created comprehensive unit tests for all four new contracts (88 tests passing)
- ✅ All test files pass code quality checks (prettier, ESLint)

### Test Update Strategy

#### Phase 1: Clean Up (✅ Complete)

- Deleted obsolete VotingAdapter test files
- Fixed remaining compilation errors in contracts

#### Phase 2: Test Infrastructure (✅ Complete)

- Created `votingConfigHelpers.ts` with helpers for:
  - Deploying ERC20/ERC721 voting configurations
  - Setting up test tokens with proper balances/ownership
  - Encoding vote data for different token types

#### Phase 3: Update Existing Tests (✅ Complete)

1. **StrategyV1.test.ts** - Updated to use VotingConfig instead of VotingAdapter ✅
2. **FreezeVotingAzoriusV1.test.ts** - Updated freeze voting to use new architecture ✅
3. **SystemDeployerV1.test.ts** - Updated deployment params for new contracts ✅
4. **StrategyV1ValidatorV1.test.ts** - Updated for new vote validation ✅

#### Phase 4: New Component Tests (✅ Complete)

1. **VotingWeightERC20V1.test.ts** - ✅ Tests weight calculation, getPastVotes, paymaster function
   - 21 tests covering initialization, weight calculation, clock modes, and ERC165 support

2. **VotingWeightERC721V1.test.ts** - ✅ Tests NFT ownership validation, duplicate checks
   - 21 tests covering initialization, NFT validation, weight calculation, and error cases

3. **VoteTrackerERC20V1.test.ts** - ✅ Tests vote recording, double-vote prevention, authorization
   - 22 tests covering initialization, vote tracking, authorization, and context handling

4. **VoteTrackerERC721V1.test.ts** - ✅ Tests token ID tracking, authorization
   - 24 tests covering NFT-specific vote tracking, preventing token reuse, and edge cases

All 88 tests are passing and code quality checks (prettier, ESLint) have been completed.

### Test Coverage Summary

The new test files thoroughly cover:
- Proxy deployment patterns with ERC1967
- Authorization patterns for vote trackers
- Clock mode handling (timestamp vs block number)
- ERC20 delegation and past votes
- NFT ownership validation and duplicate prevention
- Vote tracking per context (proposal ID)
- Interface support (ERC165)
- Edge cases and error conditions
- Large data sets (100+ NFTs)

### CODE REVIEW

File: contracts/deployables/freeze-voting/FreezeVotingBase.sol [hunk 1]
```diff
@@ -244,7 +244,8 @@ abstract contract FreezeVotingBase is
         FreezeVotingBaseStorage storage $ = _getFreezeVotingBaseStorage();
 
         // Start new freeze proposal
-        $.freezeProposalCreated = uint48(block.timestamp); // Mark creation time
+        // Use previous timestamp to ensure ERC5805 getPastVotes works
+        $.freezeProposalCreated = uint48(block.timestamp - 1); // Mark creation time
```
Comment (adamgall) [inline contracts/deployables/freeze-voting/FreezeVotingBase.sol:248]: This was a bug introduced a few PRs ago, it should have been `block.timestamp - 1` the whole time

File: contracts/deployables/freeze-voting/FreezeVotingAzoriusV1.sol [hunk 2]
```diff
@@ -282,47 +288,73 @@ contract FreezeVotingAzoriusV1 is
...
      * @param voter_ The resolved voter address
-     * @param votingAdaptersToUse_ Array of voting adapters and their data
-     * @return userVotes Total voting weight accumulated from all adapters
-     * @custom:throws InvalidVotingAdapter if adapter not in freeze proposal strategy
+     * @param votingConfigsToUse_ Array of voting configs and their data
+     * @return userVotes Total voting weight accumulated from all configs
+     * @custom:throws InvalidVotingConfig if config index is out of bounds
+     * @custom:throws NoVotingWeight if any config returns zero voting weight (includes config index and user input)
      */
     function _getVotes(
```
Comment (tbwebb22) [inline contracts/deployables/freeze-voting/FreezeVotingAzoriusV1.sol]: I know this wasn't changed in this PR - but should we find a better name for this function? getVotes sounds like a view function to me

**Review by tbwebb22 — APPROVED** [review]

File: contracts/deployables/freeze-voting/FreezeVotingAzoriusV1.sol (hunk 2 above)
Comment (adamgall) [reply contracts/deployables/freeze-voting/FreezeVotingAzoriusV1.sol]: changed to `_aggregateFreezeVotes`
